gobblin-metrics/src/main/java/gobblin/metrics/KafkaAvroReporter.java

```diff
@@ -0,0 +1,161 @@
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.Encoder;
+import org.apache.avro.io.EncoderFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Kafka reporter for codahale metrics writing metrics in Avro format
+ *
+ * @author ibuenros
+ */
+public class KafkaAvroReporter extends KafkaReporter {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(KafkaAvroReporter.class);
+
+  /**
+   * Avro schema used for encoding metrics
+   * TODO: finalize metrics avro schema
+   */
+  static String SchemaString = "{\n"+
+      " \"type\": \"record\",\n"+
+      " \"name\": \"Metric\",\n"+
+      " \"namespace\":\"<TBD>\",\n"+
+      " \"fields\" : [\n"+
+      " {\"name\": \"hostname\", \"type\": [\"null\", \"string\"], \"doc\": \"hostname emitting the metric\"},\n"+
+      " {\"name\": \"environment\", \"type\": [\"null\", \"string\"], \"doc\": \"environment where metric was emitted\"},\n"+
+      " {\"name\": \"tags\", \"type\": {\"type\": \"array\", \"items\": \"string\"}, \"doc\": \"tags associated with the metric\"},\n"+
+      " {\"name\": \"name\", \"type\": \"string\", \"doc\": \"metric name\"},\n"+
+      " {\"name\": \"value\", \"type\": [\"boolean\", \"int\", \"long\", \"float\", \"double\", \"bytes\", \"string\"], \"doc\": \"metric value\"}\n"+
+      " ]\n"+
+      "}";
+
+  Schema _schema;
+  private GenericDatumWriter<GenericRecord> w;
+  private Encoder e;
+  private ByteArrayOutputStream out;
+  private int exceptionCount;
+
+  protected KafkaAvroReporter(MetricRegistry registry, String name, MetricFilter filter, TimeUnit rateUnit,
+      TimeUnit durationUnit, String brokers, String topic, String host, String env, Set<String> tags) {
+    super(registry, name, filter, rateUnit, durationUnit, brokers, topic, host, env, tags);
+
+    Schema.Parser parser = new Schema.Parser();
+    _schema = parser.parse(SchemaString);
+
+    exceptionCount = 0;
+
+    out = new ByteArrayOutputStream();
+    e = EncoderFactory.get().binaryEncoder(out, null);
+    w = new GenericDatumWriter<GenericRecord>(_schema);
+
+  }
+
+  /**
+   * Serializes a metric key-value pair into an avro object
+   * Calls _serializeValue, which is synchronized because avro serializer is not thread-safe
+   *
+   * @param name name of the metric
+   * @param value value of the metric to report
+   * @param path additional suffixes to further identify the meaning of the reported value
+   * @return
+   */
+  @Override
+  protected byte[] serializeValue(String name, Object value, String... path) {
+    return _serializeValue(name, value, path);
+  }
+
+  protected synchronized byte[] _serializeValue(String name, Object value, String... path) {
```

Comment:
Can you simply synchronize on serializeValue()?
If you must have serializeValue() call this method, you can rename it "threadSafeSerializeValue", since we don't use underscore in method names in Gobblin.